### PR TITLE
[Blazor] Add basic ICU tests

### LINKED ISSUE #89406: [wasm] icu build related improvements
- [ ] WasmApp.targets should check that the custom filename starts with `icudt_*` since it seems to be required

- [ ] When `fullIcu==true`, then use a `nonexistant.dat` as the custom filename, so it would catch any cases where we incorrectly  use that file (it does)
https://github.com/dotnet/runtime/blob/3a88b89b025bd1f5addf61ec9a30b54284d056e2/src/mono/wasm/Wasm.Build.Tests/IcuShardingTests.cs#L242-L244

### PR DESCRIPTION
**Pull request by ilonatommy**
Fixes https://github.com/dotnet/runtime/issues/89406.
First merge https://github.com/dotnet/runtime/pull/89926, then this one.

1) Unblock Blazor WBT ICU assets checks.
2) Rename internal `LoadAllICUData` -> `LoadFullICUData`.  "All" could mean "all data files" while we meant to say "data with full locales support".
3) Fixing the problem of loading all ICU assets to runtime pack in Hybrid or Full icu mode (see `ShouldFilterCandidate`).

For a follow-up PR: https://github.com/dotnet/runtime/issues/90018

### CODE REVIEW

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+namespace Wasm.Build.Tests.Blazor;
+
+// these tests only check if correct ICU files got copied
+public class IcuTests : BlazorWasmTestBase
+{
+    public IcuTests(ITestOutputHelper output, SharedBuildPerTestClassFixture buildContext)
+        : base(output, buildContext) {}
+
+    string getRandomNameWithoutDots => Path.GetRandomFileName().Replace(".", "");
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:21]: You can use the method you added in the base class `GetRandomNameWithoutDots`.

File: src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs [hunk 2]
```diff
@@ -55,7 +55,7 @@ public abstract class ProjectProviderBase(ITestOutputHelper _testOutput, string?
         // icu
         if (assertOptions.AssertIcuAssets)
         {
-            _testOutput.WriteLine("Skipping asserting icu assets");
+            _testOutput.WriteLine("Starting icu assets assert");
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs:58]: Remove this line, and add `else { _testOutput.WriteLine("Skipping asserting icu assets"); }`, because printing `Starting .. ` is not helpful, but printing `Skipping icu .. ` could be useful in future.

File: src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/AssetsComputingHelper.cs [hunk 3]
```diff
@@ -53,7 +55,10 @@ public class AssetsComputingHelper
             ".props" when fromMonoPackage => "extension is .props is not supported.",
             ".blat" when !timezoneSupport => "timezone support is not enabled.",
             ".dat" when invariantGlobalization && fileName.StartsWith("icudt") => "invariant globalization is enabled",
+            ".dat" when loadFullICUData && fileName != "icudt" => "full ICU data is enabled",
+            ".dat" when hybridGlobalization && fileName != "icudt_hybrid" => "hybrid globalization is enabled",
             ".dat" when !string.IsNullOrEmpty(customIcuCandidateFilename) && fileName != customIcuCandidateFilename => "custom icu file will be used instead of icu from the runtime pack",
+            ".dat" when IsDefaultIcuMode() && (fileName != "icudt_EFIGS" && fileName != "icudt_CJK" && fileName != "icudt_no_CJK") => "automatic icu shard selection, based on application culture, is enabled",
```
Comment (radical) [inline src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/AssetsComputingHelper.cs:61]: Move the names of these shard files to an array, and check against that here.

File: src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/AssetsComputingHelper.cs [hunk 4]
```diff
@@ -65,6 +70,11 @@ public class AssetsComputingHelper
         };
 
         return reason != null;
+
+        bool IsDefaultIcuMode()
+        {
+            return !invariantGlobalization && !loadFullICUData && !hybridGlobalization && string.IsNullOrEmpty(customIcuCandidateFilename);
+        }
```
Comment (radical) [inline src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/AssetsComputingHelper.cs:77]: can be a statement expression:
```suggestion
        bool IsDefaultIcuMode() =>
                !invariantGlobalization &&
                !loadFullICUData &&
                !hybridGlobalization &&
                string.IsNullOrEmpty(customIcuCandidateFilename);
```

**Review by radical — COMMENTED** [review]
I'm guessing this needs a review after your other PR, that splits the tests, is merged.

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        (CommandResult res, string logPath) = BlazorBuild(
+            new BlazorBuildOptions(
+                id,
+                config,
+                WarnAsError: false,
+                GlobalizationMode: invariant == true ? GlobalizationMode.Invariant : GlobalizationMode.Hybrid,
+                ExpectedFileType: NativeFilesType.Relinked
+            ));
+        if (invariant == true)
+            Assert.Contains("$(HybridGlobalization) has no effect when $(InvariantGlobalization) is set to true.", res.Output);
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:46]: Add a `Assert.DoesNotContain` also.

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    [Theory]
+    [InlineData("Debug", false)]
+    [InlineData("Debug", true)]
+    [InlineData("Debug", null)]
+    [InlineData("Release", false)]
+    [InlineData("Release", true)]
+    [InlineData("Release", null)]
+    public async Task HybridWithInvariant(string config, bool? invariant)
+    {
+        string id = $"blz_hybrid_{config}_{GetRandomNameWithoutDots()}";
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:31]: Use the recently added `GetRandomId()` which will give you an id usable as an identifier.

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        string projectFile = CreateProjectWithNativeReference(id);
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:32]: What is the significance of the native reference here?

Comment (ilonatommy) [reply src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:32]: None, it could be just a simple tempate and actually should be

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        string extraProperties = "<HybridGlobalization>true</HybridGlobalization>";
+        if (invariant != null)
+            extraProperties += $"<InvariantGlobalization>{invariant}</InvariantGlobalization>";
+        AddItemsPropertiesToProject(projectFile, extraProperties: extraProperties);
+
+        (CommandResult res, string logPath) = BlazorBuild(
+            new BlazorBuildOptions(
+                id,
+                config,
+                WarnAsError: false,
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:42]: Why is this needed?

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    [InlineData("Debug", false)]
+    [InlineData("Debug", true)]
+    [InlineData("Debug", null)]
+    [InlineData("Release", false)]
+    [InlineData("Release", true)]
+    [InlineData("Release", null)]
+    public async Task FullIcuFromRuntimePackWithInvariant(string config, bool? invariant)
+    {
+        string id = $"blz_hybrid_{config}_{GetRandomNameWithoutDots()}";
+        string projectFile = CreateProjectWithNativeReference(id);
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:105]:
```suggestion
        string projectFile = CreateProjectWithNativeReference(id);
```

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+using Xunit.Sdk;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+#nullable enable
+
+namespace Wasm.Build.Tests.Blazor;
+
+// these tests only check if correct ICU files got copied
+public class IcuTests : BlazorWasmTestBase
```
Comment (radical) [inline src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:17]: What are the tests here testing for? Only the warning? Should they instead try to use the icu data?

Comment (ilonatommy) [reply src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:17]: They test for 1) ICU files copied to AppBundle 2) correct hierarchy of MsBuild flags 3) yes, warnings

File: src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs (hunk 8 above)
Comment (radical) [reply src/mono/wasm/Wasm.Build.Tests/Blazor/IcuTests.cs:42]: no change needed 👍

**Review by radical — APPROVED** [review]
LGTM 👍 In a follow up we should expand the tests, and maybe try to re-use the ones we have for test-main.js . I will fix up our API to facilitate that.
